### PR TITLE
[FLINK-3102] Allow reading from multiple topics with one FlinkKafkaConsumer

### PR DESCRIPTION
**Pull request by rmetzger**
* I added a new constructor to the Kafka source, accepting a List<String> of topics.
* the deserialization schema will get the topic as a String.

I'm opening the PR for review now, the tests pass locally, I'll verify the changes on a real cluster as well.

### CODE REVIEW

**Comment by aljoscha** [comment]
Did you only change `KeyedDeserializationSchema`, not `SerializationSchema`?

**Comment by rmetzger** [comment]
The `KeyedDeserializationSchema` is a new interface introduced in 1.0. I don't want to break the existing `DeserializationSchema` with the change.
Even before the change, the `KeyedDeserializationSchema` was exposing more than the DeserialziationSchema (the offset), it can be seen as the Rich* variant of the DeserSchema.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Fetcher.java [hunk 1]
```diff
@@ -61,15 +53,15 @@
...
-	 * 
+	 *
+	 * @param <T> The type of elements produced by the fetcher and emitted to the source context.
 	 * @param sourceContext The source context to emit elements to.
 	 * @param valueDeserializer The deserializer to decode the raw values with.
-	 * @param lastOffsets The array into which to store the offsets for which elements are emitted (operator state)
-	 * 
-	 * @param <T> The type of elements produced by the fetcher and emitted to the source context.
+	 * @param lastOffsets The map into which to store the offsets for which elements are emitted (operator state)
+*
```
Comment (aljoscha) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Fetcher.java]: Strange formatting of `*`. Did you check whether the generic param should be before normal params or why did you move?

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (aljoscha) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java]: The file has inconsistent control flow style, some of the `if` have a space before opening parenthesis the `for` loops don't have a space.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartitionLeader.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
+/*
```
Comment (aljoscha) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartitionLeader.java]: Inconsistent control flow parenthesis style.

**Comment by aljoscha** [comment]
The style of control-flow statements is not consistent, some have space before opening parenthesis and some don't.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 4]
```diff
@@ -516,10 +510,12 @@ public void close() throws Exception {
...
 		if (LOG.isDebugEnabled()) {
 			LOG.debug("Snapshotting state. Offsets: {}, checkpoint id: {}, timestamp: {}",
-					Arrays.toString(lastOffsets), checkpointId, checkpointTimestamp);
+					KafkaTopicPartition.toString(lastOffsets), checkpointId, checkpointTimestamp);
 		}
 
-		long[] currentOffsets = Arrays.copyOf(lastOffsets, lastOffsets.length);
+		// the use of clone() is okay here is okay, we just need a new map, the keys are not changed
+		//noinspection unchecked
+		HashMap<KafkaTopicPartition, Long> currentOffsets = (HashMap<KafkaTopicPartition, Long>) lastOffsets.clone();
```
Comment (aljoscha) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Why don't you use `new HashMap<>(lastOffsets)` here?

**Comment by aljoscha** [comment]
Aside from minor comments and the style issue it looks good. I'm not too familiar with the internals of Kafka (which we are touching here, a bit), though.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 5]
```diff
@@ -25,6 +25,7 @@
 import kafka.javaapi.consumer.SimpleConsumer;
 import org.apache.commons.collections.map.LinkedMap;
 
+import org.apache.commons.lang.StringUtils;
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Can we consistently use the classes from "lang3" instread of "lang" v2?

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 6]
```diff
@@ -303,25 +300,29 @@ public FlinkKafkaConsumer(String topic, KeyedDeserializationSchema<T> deserializ
...
 
-		// get initial partitions list. The order of the partitions is important for consistent 
-		// partition id assignment in restart cases.
-		this.partitions = new int[partitionInfos.size()];
-		for (int i = 0; i < partitionInfos.size(); i++) {
-			partitions[i] = partitionInfos.get(i).partition();
-			
-			if (partitions[i] >= partitions.length) {
-				throw new RuntimeException("Kafka partition numbers are sparse");
+		Map<String, Integer> countPerTopic = new HashMap<>();
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: How about guarding this section with `if (LOG.isInfoEnabled())`?

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java (hunk 4 above)
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Clone is probably a bit more efficient even, as does not hash but only copy the table.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 7]
```diff
@@ -372,34 +372,28 @@ public void open(Configuration parameters) throws Exception {
...
-					lastOffsets[i] = restoredOffset;
-				}
+			for(Map.Entry<KafkaTopicPartition, Long> restorePartition: restoreToOffset.entrySet()) {
+				// seek fetcher to restore position
+				// we set the offset +1 here, because seek() is accepting the next offset to read,
+				// but the restore offset is the last read offset
+				fetcher.seek(restorePartition.getKey(), restorePartition.getValue() + 1);
 			}
+			// initialize offsets with restored state
+			this.lastOffsets = restoreToOffset;
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: For GC friendlyness, how about setting `restoreToOffset = null;`

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java (hunk 5 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: I will exclude all lang imports using a checkstyle rule.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 8]
```diff
@@ -563,37 +559,44 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
...
 				
 				// remove older checkpoints in map
 				for (int i = 0; i < posInMap; i++) {
 					pendingCheckpoints.remove(0);
 				}
 			}
-	
-			if (LOG.isInfoEnabled()) {
-				LOG.info("Committing offsets {} to offset store: {}", Arrays.toString(checkpointOffsets), offsetStore);
+			if(checkpointOffsets.size() == 0) {
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: I think this is a bug. If the list is empty after you remove the checkpoint to commit, it only means that no pending checkpoint was subsumed. It should still commit the offsets to Kafka.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 9]
```diff
@@ -222,18 +219,18 @@
...
+	private transient List<KafkaTopicPartitionLeader> subscribedPartitions;
 
 	/** The offsets of the last returned elements */
-	private transient long[] lastOffsets;
+	private transient HashMap<KafkaTopicPartition, Long> lastOffsets;
 
 	/** The latest offsets that have been committed to Kafka or ZooKeeper. These are never
 	 * newer then the last offsets (Flink's internal view is fresher) */
-	private transient long[] commitedOffsets;
+	private transient HashMap<KafkaTopicPartition, Long> commitedOffsets;
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Type in the variable name: "committed"

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 10]
```diff
@@ -563,37 +559,44 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
...
+				Long offset = checkpointOffsets.get(tp);
+				Long lastCommitted = commitedOffsets.get(tp);
+				if(lastCommitted == null) {
+					lastCommitted = OFFSET_NOT_SET;
+				}
 				if (offset != OFFSET_NOT_SET) {
 					if (offset > lastCommitted) {
 						offsetsToCommit.put(tp, offset);
-						LOG.debug("Committing offset {} for partition {}", offset, partition);
+						//noinspection unchecked
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Inspection suppression should not be needed...

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 11]
```diff
@@ -563,37 +559,44 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
...
+						commitedOffsets.put(tp, offset);
+						LOG.debug("Committing offset {} for {}", offset, tp.toString());
 					}
 					else {
-						LOG.debug("Ignoring offset {} for partition {} because it is already committed", offset, partition);
+						LOG.debug("Ignoring offset {} for partition {} because it is already committed", offset, tp);
 					}
 				}
 			}
+			if(LOG.isInfoEnabled() && offsetsToCommit.size() > 0) {
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Is this a good thing on INFO level? Will be a lot of contents in the logs, especially when running frequent checkpoints. I would vote to log this on DEBUG...

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Fetcher.java (hunk 1 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Fetcher.java]: Fixed the `*`.
The java standard library is putting generic parameters first.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 12]
```diff
@@ -640,30 +642,38 @@ public PeriodicOffsetCommitter(long commitInterval, FlinkKafkaConsumer consumer)
...
 						// create copy of current offsets
-						long[] currentOffsets = Arrays.copyOf(consumer.lastOffsets, consumer.lastOffsets.length);
-
-						Map<TopicPartition, Long> offsetsToCommit = new HashMap<>();
-						for (TopicPartition tp : (List<TopicPartition>)consumer.subscribedPartitions) {
-							int partition = tp.partition();
-							long offset = currentOffsets[partition];
-							long lastCommitted = consumer.commitedOffsets[partition];
-
+						HashMap<KafkaTopicPartition, Long> currentOffsets = (HashMap<KafkaTopicPartition, Long>) consumer.lastOffsets.clone();
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Code duplication, can go into a util method

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 13]
```diff
@@ -640,30 +642,38 @@ public PeriodicOffsetCommitter(long commitInterval, FlinkKafkaConsumer consumer)
...
+									consumer.commitedOffsets.put(tp, offset);
+									LOG.debug("Committing offset {} for partition {}", offset, tp);
 								} else {
-									LOG.debug("Ignoring offset {} for partition {} because it is already committed", offset, partition);
+									LOG.debug("Ignoring offset {} for partition {} because it is already committed", offset, tp);
 								}
 							}
 						}
 
+						if(LOG.isInfoEnabled() && offsetsToCommit.size() > 0) {
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Same as above: Info logging may be pretty aggressive...

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A serializable representation of a kafka topic and a partition.
+ * Used as an operator state for the Kafka consumer
+ */
+public class KafkaTopicPartition implements Serializable {
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java]: Can this be an immutable type?

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java [hunk 15]
```diff
@@ -640,30 +642,38 @@ public PeriodicOffsetCommitter(long commitInterval, FlinkKafkaConsumer consumer)
...
-						Map<TopicPartition, Long> offsetsToCommit = new HashMap<>();
-						for (TopicPartition tp : (List<TopicPartition>)consumer.subscribedPartitions) {
-							int partition = tp.partition();
-							long offset = currentOffsets[partition];
-							long lastCommitted = consumer.commitedOffsets[partition];
-
+						HashMap<KafkaTopicPartition, Long> currentOffsets = (HashMap<KafkaTopicPartition, Long>) consumer.lastOffsets.clone();
+
+						Map<KafkaTopicPartition, Long> offsetsToCommit = new HashMap<>();
+						for (KafkaTopicPartition tp : (List<KafkaTopicPartition>)consumer.subscribedPartitions) {
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: A bit unclean to cast here from `List<KafkaTopicPartitionLeader>` to `List<KafkaTopicPartition>`. Why not simply iterate with the `KafkaTopicPartitionLeader` type?

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java (hunk 14 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java]: Good catch!
I'll change it.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartitionLeader.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+import java.io.Serializable;
+import java.util.List;
+
+/**
+ * Serializable Topic Partition info with leader Node information.
+ * This class is used at runtime.
+ *
+ * The leader information does not account for equality!
+ */
+public class KafkaTopicPartitionLeader extends KafkaTopicPartition implements Serializable {
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartitionLeader.java]: This should be an immutable type as well.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java (hunk 15 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer.java]: Fixed. I forgot to set the type parameter for the `consumer` reference, that's why I did excessive casing here.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+				", partition=" + partition +
+				'}';
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || !(o instanceof KafkaTopicPartition)) {
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java]: Redundant, "instanceof" checks for null.

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+			return false;
+		}
+
+		KafkaTopicPartition that = (KafkaTopicPartition) o;
+
+		if (partition != that.partition) {
+			return false;
+		}
+		return !(topic != null ? !topic.equals(that.topic) : that.topic != null);
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartition.java]: Make topic a "non null" fields (check in the constructor) and save a few cycles here...

File: flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartitionLeader.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+		StringBuilder sb = new StringBuilder();
+		for(KafkaTopicPartitionLeader p: partitions) {
+			sb.append(p.getTopic()).append(":").append(p.getPartition()).append(", ");
+		}
+		return sb.toString();
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		return super.equals(o);
```
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/KafkaTopicPartitionLeader.java]: This is ignoring the leader IDs. Seems wrong...

**Comment by StephanEwen** [comment]
All in all quite good. Few things:

  - The code style is super heterogeneous, sometimes it puts spaces after a keyword, sometimes it does not. Would be good to make this straight.

  - This mixes "Arrays.toString" and "StringUtils.join". Why not avail any dependency on external libraries (with all the dependency conflicts, it is nice to avoid where possible) and use
     - `Arrays.toString()` for arrays.
     - Simply `toString()` for lists.

  - `KafkaTopicPartition` and `KafkaTopicPartitionLeader` need to be fast for hash code and equals. They are used as hash map keys for every record (offset update). Due to that, they should really be immutable types and cache their hash code, since that is used for any record emitted (update in the offsets hash map). Also, equals() should be fast.

**Comment by rmetzger** [comment]
Thank you for the good reviews!
I've addressed all comments.
The entire kafka connector is now using a space after `if` and `for`.

**Comment by StephanEwen** [comment]
Looks pretty good.

+1 to merge
